web/middlewares/secure.go

```diff
@@ -0,0 +1,112 @@
+package middlewares
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/instance"
+	"github.com/labstack/echo"
+)
+
+type (
+	XFrameOption string
+	CSPSource    int
+
+	// SecureConfig defines the config for Secure middleware.
+	SecureConfig struct {
+		HSTSMaxAge    time.Duration
+		CSPScriptSrc  []CSPSource
+		CSPFrameSrc   []CSPSource
+		CSPConnectSrc []CSPSource
+		XFrameOptions XFrameOption
+		XFrameAllowed string
+	}
+)
+
+const (
+	XFrameDeny       XFrameOption = "DENY"
+	XFrameSameOrigin              = "SAMEORIGIN"
+	XFrameAllowFrom               = "ALLOW-FROM"
+
+	CSPSrcSelf CSPSource = iota
+	CSPSrcParent
+	CSPSrcParentSubdomains
+	CSPSrcAny
+)
+
+func Secure(conf *SecureConfig) echo.MiddlewareFunc {
+	var hstsHeader string
+	if conf.HSTSMaxAge > 0 {
+		hstsHeader = fmt.Sprintf("max-age=%.f; includeSubdomains",
+			conf.HSTSMaxAge.Seconds())
+	}
+
+	var xFrameHeader string
+	switch conf.XFrameOptions {
+	case XFrameDeny:
+		xFrameHeader = string(XFrameDeny)
+	case XFrameSameOrigin:
+		xFrameHeader = string(XFrameSameOrigin)
+	case XFrameAllowFrom:
+		xFrameHeader = fmt.Sprintf("%s %s", XFrameAllowFrom, conf.XFrameAllowed)
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			hsts := true
+			if in := c.Get("instance"); in != nil && in.(*instance.Instance).Dev {
+				hsts = false
```

Comment:
I'm not sure we should not also disable CSP for dev instances. I wonder if it will interfer with debugging tools. Maybe we can start like that and see if it causes trouble.